From 4930140ec6fdceb7e15d31afa45243dee7d1e217 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Thu, 31 Aug 2017 16:03:06 +0000
Subject: [PATCH] [CONFIGURATION-671] Fixed handling of empty sections in
 INIConfiguration.

Saving of an INIConfiguration with an empty section caused a NPE. The
cause was that nodes without children were not detected as section
nodes. Now a different criterion is used.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1806819 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration2/INIConfiguration.java      |  6 +++++-
 .../configuration2/TestINIConfiguration.java  | 21 +++++++++++++++++++
 2 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/configuration2/INIConfiguration.java b/src/main/java/org/apache/commons/configuration2/INIConfiguration.java
index 9063ec5096..c9f516ab96 100644
--- a/src/main/java/org/apache/commons/configuration2/INIConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/INIConfiguration.java
@@ -212,6 +212,10 @@
  * <a href="http://commons.apache.org/proper/commons-configuration/userguide/howto_basicfeatures.html">
  * Basic features and AbstractConfiguration</a> of the user's guide.
  * </p>
+ * <p>
+ * Note that this configuration does not support properties with null values.
+ * Such properties are considered to be section nodes.
+ * </p>
  *
  * @author <a
  *         href="http://commons.apache.org/configuration/team-list.html">Commons
@@ -942,7 +946,7 @@ private SubnodeConfiguration getGlobalSection()
      */
     private static boolean isSectionNode(ImmutableNode node)
     {
-        return !node.getChildren().isEmpty();
+        return node.getValue() == null;
     }
 
     /**
diff --git a/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java
index 02924fc75f..30b8722610 100644
--- a/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestINIConfiguration.java
@@ -1223,6 +1223,27 @@ public void testExpressionEngineIgnoringCase()
                 config.getString("SECTION1.VAR1"));
     }
 
+    /**
+     * Tests whether an empty section can be saved. This is related to
+     * CONFIGURATION-671.
+     */
+    @Test
+    public void testWriteEmptySection()
+            throws ConfigurationException, IOException
+    {
+        final String section = "[EmptySection]";
+        INIConfiguration config = setUpConfig(section);
+        assertEquals("Wrong number of sections", 1,
+                config.getSections().size());
+        assertTrue("Section not found",
+                config.getSections().contains("EmptySection"));
+
+        StringWriter writer = new StringWriter();
+        config.write(writer);
+        assertEquals("Wrong saved configuration",
+                section + LINE_SEPARATOR + LINE_SEPARATOR, writer.toString());
+    }
+
     /**
      * A thread class for testing concurrent access to the global section.
      */
